Source/Tools/LinuxEmulation/LinuxSyscalls/LinuxAllocator.cpp

```diff
@@ -515,7 +515,19 @@ uint64_t MemAllocator32Bit::Shmdt(const void* shmaddr) {
   }
 
   uint64_t Result = ::shmdt(shmaddr);
-  PageToShm.erase(it);
+
+  if (Result != -1) {
+    int shmid = it->second;
+    PageToShm.erase(it);
+
+    struct shmid_ds buf {};
+    if (shmctl(shmid, IPC_STAT, &buf) == 0) {
```

Comment:
I think this needs to occur before the `shmdt` because if the refcount of the sysv shm region `shm_nattch` reaches zero (which shmdt decrements) then it gets deleted before we can get the information.

Reply:
Alternative take `PageToShm[LowerPage] = shmid;` in the above `shmat` handler should store ID and size, so we can recover it without doing a shmctl here.

Reply:
Oh right, fixed. Also moved the `erase` call inside an if, to prevent the rare occasion that it could mess the shmdt errno.